lib/cretonne/src/isa/intel/mod.rs

```diff
@@ -111,4 +117,149 @@ impl TargetIsa for Isa {
+        for _reg in &csrs {
+            csr_stack_size += word_size;
+        }
+
+        let stack_offset = -(csr_stack_size as i32);
+        let slot = ir::StackSlotData {
+            kind: ir::StackSlotKind::IncomingArg,
+            size: csr_stack_size,
+            offset: stack_offset,
+        };
+        func.create_stack_slot(slot);
+
+        let total_stack_size = layout_stack(&mut func.stack_slots, word_size)?;
+        let local_stack_size = total_stack_size - csr_stack_size;
+
+        // Build up list of args, which we'll append forwards to the params and
+        // backwards to the returns.
+        let mut csr_args = Vec::new();
+        csr_args.push(ir::AbiParam::special_reg(
+            csr_type,
+            ir::ArgumentPurpose::FramePointer,
+            RU::rbp as RegUnit,
+        ));
+        for reg in &csrs {
+            csr_args.push(ir::AbiParam::special_reg(
+                csr_type,
+                ir::ArgumentPurpose::CalleeSaved,
+                *reg as RegUnit,
+            ));
+        }
+
+        for csr_arg in &csr_args {
+            func.signature.params.push(*csr_arg);
+        }
+        for csr_arg in csr_args.iter().rev() {
+            func.signature.returns.push(*csr_arg);
+        }
+
+        // Append param to entry EBB
+        let entry_ebb = func.layout.entry_block().expect("missing entry block");
+        func.dfg.append_ebb_param(entry_ebb, csr_type);
+
+        // Find our frame pointer parameter Value
+        let fp = func.special_param(ir::ArgumentPurpose::FramePointer)
+            .expect("missing frame pointer");
+
+        // Assign it a location
+        func.locations[fp] = ir::ValueLoc::Reg(RU::rbp as RegUnit);
+
+        let mut csr_vals = Vec::new();
+        for reg in &csrs {
+            // Append param to entry EBB
+            func.dfg.append_ebb_param(entry_ebb, csr_type);
+
+            let csr_arg = func.dfg.ebb_params(entry_ebb).last().expect(
+                "no last argument",
+            );
+
+            // Assign it a location
+            func.locations[*csr_arg] = ir::ValueLoc::Reg(*reg as RegUnit);
+
+            // Remember it so we can push it momentarily
+            csr_vals.push(*csr_arg);
+        }
+
+
+        // Insert prologue
+        {
+            let mut pos = EncCursor::new(func, self).at_first_insertion_point(entry_ebb);
+            pos.ins().x86_push(fp);
+            pos.ins().copy_special(
+                RU::rsp as RegUnit,
+                RU::rbp as RegUnit,
+            );
+            if local_stack_size > 0 {
+                pos.ins().adjust_sp_imm(-(local_stack_size as i32));
+            }
+
+            for csr_arg in csr_vals {
+                pos.ins().x86_push(csr_arg);
+            }
+        }
+
+        // Find all 'return' instructions
+        let mut return_insts = Vec::new();
+        for ebb in func.layout.ebbs() {
+            for inst in func.layout.ebb_insts(ebb) {
```

Comment:
We don't have any conditional return instructions, and I don't think Intel ever will, so return instructions are always going to be [terminator instructions](http://cretonne.readthedocs.io/en/latest/langref.html#term-terminator-instruction). This means they are guaranteed to be at the end of an EBB, so you can just check the last instruction.

For an algorithm like this, I would use a cursor to navigate the EBBs and  pass the cursor instead of the function into `insert_epilogue` for mutation.